vxvas2nets/vas2nets_ussd.py

```diff
@@ -0,0 +1,223 @@
+
+
+class Vas2NetsUssdTransport(HttpRpcTransport):
+    CONFIG_CLASS = Vas2NetsUssdTransportConfig
+    EXPECTED_FIELDS = frozenset([
+        'userdata', 'endofsession', 'msisdn', 'sessionid',
+    ])
+
+    transport_type = 'ussd'
+
+    @inlineCallbacks
+    def setup_transport(self):
+        yield super(Vas2NetsUssdTransport, self).setup_transport()
+        config = self.get_static_config()
+        r_prefix = "vumi.transports.vas2nets_ussd:%s" % self.transport_name
+        self.session_manager = yield SessionManager.from_redis_config(
+            config.redis_manager, r_prefix,
+            max_session_length=config.ussd_session_timeout)
+        self.ussd_number = config.ussd_number
+
+    def get_request_dict(self, request):
+        return {
+            'uri': request.uri,
+            'method': request.method,
+            'path': request.path,
+            'content': request.content.read(),
+            'headers': dict(request.requestHeaders.getAllRawHeaders()),
+        }
+
+    @inlineCallbacks
+    def session_event_for_transaction(self, session_id, endofsession):
+        if endofsession == 'true':
+            yield self.session_manager.clear_session(session_id)
+            returnValue(TransportUserMessage.SESSION_CLOSE)
+
+        session = yield self.session_manager.load_session(session_id)
+        if session:
+            yield self.session_manager.save_session(session_id, session)
+            returnValue(TransportUserMessage.SESSION_RESUME)
+        else:
+            yield self.session_manager.create_session(
+                session_id, transaction_id=session_id)
+            returnValue(TransportUserMessage.SESSION_NEW)
+
+    @inlineCallbacks
+    def handle_raw_inbound_message(self, message_id, request):
+        try:
+            values, errors = self.get_field_values(
+                request, self.EXPECTED_FIELDS)
+        except UnicodeDecodeError:
+            self.log.msg('Bad request encoding: %r' % request)
+            request_dict = self.get_request_dict(request)
+            self.finish_request(
+                message_id, json.dumps({'invalid_request': request_dict}),
+                code=http.BAD_REQUEST)
+            yield self.add_status(
+                component='request', status='down', type='invalid_encoding',
+                message='Invalid encoding', details={'request': request_dict})
+            return
+
+        if errors:
+            self.log.msg('Unhappy incoming message: %s ' % (errors,))
+            yield self.finish_request(
+                message_id, json.dumps(errors), code=http.BAD_REQUEST
+            )
+            yield self.add_status(
+                component='request', status='down',
+                type='invalid_inbound_fields',
+                message='Invalid inbound fields', details=errors)
+            return
+
+        yield self.add_status(
+            component='request', status='ok', type='request_parsed',
+            message='Request parsed',)
```

Comment:
I'm wondering about the interaction between this status ok and the slow response down time statuses. If slow responses are coming in, that's mean the status would flip-flop between ok and down or degraded on every message. Perhaps we should leave returning the status to ok for when the response comes back?